submissions/kt-std/object-oriented-js/app.js

```diff
@@ -0,0 +1,255 @@
+    ROW_NUM = 6,
+    COL_NUM = 7,
+    ROW_HEIGHT = 70,
+    Y_STEP = 70,
+    X_STEP = 70,
+    INITIAL_X = X_STEP * Math.floor(COL_NUM / 2),
+    INITIAL_Y = (ROW_NUM - 1) * Y_STEP,
+    CANVAS_START = 0,
+    CANVAS_WIDTH = COL_NUM * X_STEP,
+    CANVAS_HEIGHT = ROW_NUM * X_STEP + 20,
+    LIVES_AMOUNT = 3,
+    LIVE_SPRITE = "images/heart.svg",
+    LIVES_CONTAINER = document.getElementById("hearts"),
+    RANDOM_BASE = 100,
+    SPEED_INC = 0.01,
+    X_GAP = X_STEP / 2,
+    Y_GAP = Y_STEP / 2,
+    messages = {
+        win: {
+            text: `You've just successfully saved Xmas!`,
+            path: `images/bell.svg`,
+            button: `Play again!`,
+            heading: `Congrats!`,
+        },
+        lose: {
+            text: ``,
+            path: `images/bell.svg`,
+            button: `Try again!`,
+            heading: `You lose \u26C4`,
+        },
+    };
+
+const Enemy = function (index) {
+    this.x = CANVAS_START;
+    this.y = Y_STEP * index;
+    this.speed = Math.random() * RANDOM_BASE;
+    this.sprite = "images/yeti.png";
+};
+
+/* This function displays each Enemy instance sprite on the canvas
+ * due to its current X and Y coodinates
+ */
+Enemy.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.update = function (dt) {
+    // Speed up the enemy during the game
+    this.speed += SPEED_INC;
+    // Update Enemy's horizontal position
+    this.x = this.updateHorizontalPosition(this.x + dt * this.speed);
+};
+
+Enemy.prototype.updateHorizontalPosition = function (position) {
+    /* If the Enemy moves right off the visible canvas area
+     * place it back to the beginning of the row in other cases
+     * move right
+     */
+    return position > CANVAS_WIDTH + X_GAP ? CANVAS_START - X_GAP : position;
+};
+
+const Player = function () {
+    this.y = INITIAL_Y;
+    this.x = INITIAL_X;
+    this.lives = LIVES_AMOUNT;
+    this.sprite = "images/santa.png";
+};
+
+/* This function displays Player sprite on the canvas
+ * due to its current X and Y coodinates
+ */
+Player.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+/* This function checks if there are left any Player lives.
+ * If there remain no lives, display a loss message.
+ * If Player still left lives and it has crossed the border of any Enemy,
+ * then decrease Player lives amount, delete live item from the LIVES_CONTAINER
+ * and reset Player position to the initial state.
+ */
+Player.prototype.update = function () {
+    if (!this.lives) {
+        displayMessage("lose", "flex", messages);
+    } else {
+        allEnemies.forEach((enemy) => {
+            if (
+                Math.abs(this.x - enemy.x) < X_GAP &&
+                Math.abs(this.y - enemy.y) < Y_GAP
+            ) {
+                this.lives--;
+                removeLivesFromContainer();
+                resetPosition(this, INITIAL_X, INITIAL_Y);
```

Comment:
`
this.setPosition(INITIAL_X, INITIAL_Y);
`

what about this instead of global func resetPosition()?

Reply:
Not annoying at all! :)
Thanks a lot for your detailed review.